gateway/ln-gateway/src/lightning/ldk.rs

```diff
@@ -290,7 +356,50 @@ impl ILnRpcClient for GatewayLdkClient {
         max_delay: u64,
         max_fee: Amount,
     ) -> Result<PayInvoiceResponse, LightningRpcError> {
-        let payment_id = match self.node.bolt11_payment().send(
+        // Subscribe to the `PaymentSuccessful`/`PaymentFailed` events for this invoice.
+        let (sender, receiver) = tokio::sync::oneshot::channel();
+        self.payment_subscriptions
+            .lock()
+            .await
+            .insert(*invoice.payment_hash(), sender);
```

Comment:
You raise a good point: This should handle multiple concurrent calls while the payment is in-flight

I changed `PaymentSubscriptions` to map to a _list_ of oneshot senders

Reply:
Using a list here feels wrong. I think its just due to some implementation details, but let me know if I'm missing something more fundamental here. Can you do something like this:

1) Lock `payment_subscriptions` map
2) Check if the payment has already been made. You already do this in `pay_idempotent`
3) Only if it has not been made, then insert the sender into the map
4) Do payment as you do now

This would require some refactoring to pull out 2) from `pay_idempotent`, but I think it allows you to assert that there is only ever 1 thread waiting on a oneshot channel.